File: sorting/mergesort.java
public class Main{
	public static void main(String args[]){
	    int A[] = {23,1,45,12,6,5,43,76,14,90,87};
		mergesort(A);
		for(int i =0 ; i < A.length; i++){
			System.out.print(A[i]+" ");
		}
	}

	public static void mergesort(int A[]){

		int length = A.length;
		if(length <= 1) return;
		int mid = A.length/2;
		int left[] = new int[mid];
		int right[] = new int[length - mid];
		for(int i = 0; i < mid; i++){
			left[i] = A[i];
		}
		for(int j = mid; j < length; j++){
			right[j-mid] = A[j];
		}
		mergesort(left);
		mergesort(right);
		merge(left,right,A);

	}
	public static void merge(int[] left, int[] right, int[] A){

		int i = 0, j = 0, k =0;
		
		while(i < left.length && j < right.length){
			if(left[i] < right[j]){
				A[k] = left[i];
				i++;
			}else{
				A[k] = right[j];
				j++;
			}
			
			k++;
		}
		while(i < left.length){
			A[k] = left[i];
            i++;
            k++;
		}

		while(j < right.length){
			A[k] = right[j];
            j++;
            k++;
		}
	}
}

File: algortihms/mergesort.cpp
#include <algorithm>
#include<iostream>
#include <vector>

using namespace std;

vector<int> merge_sort(vector<int>& vec, int l , int r){
    vector<int> merged;
    if(l == r){
        merged.push_back(vec[l]);
        return  merged;
    }

    int mid = (l+r)/2;

    vector<int> left = merge_sort(vec,l,mid);
    vector<int> right = merge_sort(vec,mid+1,r);


    int left_size = left.size();
    int right_size = right.size();

    int i = 0;
    int j = 0;

    while(i < left_size && j < right_size){
        if(left[i] < right[j]){
            merged.push_back(left[i]);
            i++;
        }else if(left[i] > right[j]){
           merged.push_back(right[j]);
           j++;
        }else{
            merged.push_back(left[i]);
            merged.push_back(right[j]);
            i++;
            j++;
        }
    }

    while(i < left_size){
        merged.push_back(left[i]);
        i++;
    }

    while(j < right_size){
        merged.push_back(right[j]);
        j++;
    }

    return merged;
}


int main(){

    vector<int> vec = {4,1,9,45,12,15,0,-4,2,7};
    for(int i = 0; i < 10; i++){
        vec[i] = rand()%100;
    }

    for(int i = 0; i < 10; i++){
        cout << vec[i] << ",";
    }

    cout << endl;

    vector<int> sorted_arr = merge_sort(vec, 0, vec.size()-1);

    for(int i = 0; i < 10; i++){
        cout << sorted_arr[i] << ",";
    }
    sort(vec.begin(), vec.end());

    bool flag = true;

    for(int i = 0; i < 10; i++){
        if(vec[i] != sorted_arr[i]){
            flag = false;
        }
    }

    cout << endl;
    if(flag){
        cout << "ACCEPTED !";
    }else{
        cout << "FAILED !";
    }

    return 0;
}
